src/targets/gpu/mlir.cpp

```diff
@@ -167,6 +168,51 @@ std::string mlir_print(F f, T x)
     return ss.str();
 }
 
+struct mlir_logger
+{
+    std::string diagnostics;
+    std::stringstream ss;
+
+    mlir_context& ctx;
+    MlirDiagnosticHandlerID id;
+
+    mlir_logger(mlir_context& ctx);
+    ~mlir_logger();
```

Comment:
Add a default constructor, move constructor and move assignment operators and delete the copy constructor and copy assignment.